### DISCUSSION TITLE
The final variable of TableData, how to use the portable setter

### DISCUSSION THREAD

**Discussion by greenking19** [simolus3/drift#1897] [General]
👇 **Example**
```dart
TaskTableData _taskData = await appDatabase.select(appDatabase.taskTable).getSingle();
 ///  change property
 ///  error: Try finding a different setter, or making 'taskId' non-final.
 _taskData.taskId = 'test';

 /// Insert/Upgrade Table
await appDatabase.into(appDatabase.taskTable).insertOnConflictUpdate(_taskData);
```
This is wrong, are there other alternatives?
Now I use TableData.toJson and TableData.formJson, then insertOnConflictUpdate

---------------------------
I also encountered a small problem when developing
If there is no such data in the table, it will prompt a null error
```dart
TaskTableData _list = await (appDatabase.select(appDatabase.taskTable)
          ..where((tbl) => tbl.taskId.equals(taskId)))
        .getSingle();
```
But if I write, editor prompt always true
<img width="807" alt="image" src="https://user-images.githubusercontent.com/43332372/174662547-c4203727-c71b-4e0e-8ba8-dd8325f02d21.png">

In the end, I wrote it like this.
<img width="861" alt="image" src="https://user-images.githubusercontent.com/43332372/174663071-90e0c79c-dab1-4545-aa73-8286c8a7ae55.png">

**Comment by greenking19** [comment]
🤯 Damn, I forgot to have TableCompanion.
🤯 Sorry I made this basic mistake.

-----------
Currently I only have this problem with **getSingle**. 
When the table doesn't have the data I want it throws a null error

**Comment by simolus3** [comment]
> This is wrong, are there other alternatives?

You _could_ instruct drift to generate mutable data classes with the `mutable_classes` [build option](https://drift.simonbinder.eu/docs/advanced-features/builder_options/). But that's a pretty major switch and recommend keeping them immutable, as that's the way most drift APIs were designed.

I'd just use `_taskData = _taskData.copyWith(taskId: 'test')` in this case (the same works for companions).

> Currently I only have this problem with __getSingle__.

Just use `getSingleOrNull()`, it's exactly what you need here! `getSingle` is written under the assumption that you're going to have at least one row.

**Reply by greenking19** [reply]
Thank you for clearing my doubts.
😋

**Comment by greenking19** [comment]
I want to use TableData in the modified form. It needs to be populated during initialization.
But I found that toJson is not what I expected
```dart
class TestTable extends Table{
  TextColumn get id => text().named('ID')();
  TextColumn get message text().named('Message')();
  @override
  Set<Column> get primaryKey => {id};
}
```
this my conversion class
```dart
class TestConvert {
  String? iD;
  String? message;

  TestConvert({this.iD, this.message});

  TestConvert.fromJson(Map<String, dynamic> json) {
    iD = json['ID'];
    message = json['Message'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['ID'] = this.iD;
    data['Message'] = this.message;
    return data;
  }
}
```
my expectations.
Query out the old data, and then fill it into ConvertClass. Then hand it over to the form to Change
then save
```dart
// A testTableData that was queried
TestTableData testTable = xxx.get();

TestConvert formData = TestConvert.formJson(testTable.toJson());
```
But I found that the generated TableData.toJson is like this.
I have a lot of forms. They are all for inserting into the database. TableData cannot be used directly.
Currently I can only modify my ConvertClass. Is there any other solution?
```dart
 Map<String, dynamic> toJson({ValueSerializer? serializer}) {
    serializer ??= driftRuntimeOptions.defaultSerializer;
    return <String, dynamic>{
      'id': serializer.toJson<String>(id),
      'message': serializer.toJson<String>(message),
   );
}
```

**Reply by simolus3** [reply]
`named` only affects the name of the column in SQL. You need to add a `@JsonKey` annotation onto the column to change the JSON key.

But if your intention is to go from drift data classes from and to your form data classes, it's much easier to just make drift use your form classes directly! You can do this with [custom data classes](https://drift.simonbinder.eu/docs/advanced-features/custom_row_classes/):

```dart
@UseRowClass(TestConvert)
class TestTable extends Table{
  TextColumn get id => text()();
  TextColumn get message text()();
  @override
  Set<Column> get primaryKey => {id};
}
```

**Reply by greenking19** [reply]
Thank you now the problem is solved. This is my problem, I don't have the documentation.
thanks for your help
